File: frameanimation/src/main/java/com/github/penfeizhou/animation/FrameAnimationDrawable.java
package com.github.penfeizhou.animation;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.DrawFilter;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.PaintFlagsDrawFilter;
import android.graphics.PixelFormat;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.vectordrawable.graphics.drawable.Animatable2Compat;

import com.github.penfeizhou.animation.decode.BaseFrameSeqDecoder;
import com.github.penfeizhou.animation.decode.FrameSeqDecoder2;
import com.github.penfeizhou.animation.loader.Loader;

import java.lang.ref.WeakReference;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public abstract class FrameAnimationDrawable<Decoder extends FrameSeqDecoder2<?>>
        extends Drawable implements Animatable2Compat, BaseFrameSeqDecoder.RenderListener {
    private static final String TAG = FrameAnimationDrawable.class.getSimpleName();
    private final Paint paint = new Paint();
    private final Decoder frameSeqDecoder;
    private final DrawFilter drawFilter = new PaintFlagsDrawFilter(0, Paint.ANTI_ALIAS_FLAG | Paint.FILTER_BITMAP_FLAG);
    private final Matrix matrix = new Matrix();
    private final Set<AnimationCallback> animationCallbacks = new HashSet<>();
    private Bitmap bitmap;
    private static final int MSG_ANIMATION_START = 1;
    private static final int MSG_ANIMATION_END = 2;
    private final Handler uiHandler = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case MSG_ANIMATION_START: {
                    ArrayList<AnimationCallback> callbacks = new ArrayList<>(animationCallbacks);
                    for (AnimationCallback animationCallback : callbacks) {
                        animationCallback.onAnimationStart(FrameAnimationDrawable.this);
                    }
                    break;
                }
                case MSG_ANIMATION_END: {
                    ArrayList<AnimationCallback> callbacks = new ArrayList<>(animationCallbacks);
                    for (AnimationCallback animationCallback : callbacks) {
                        animationCallback.onAnimationEnd(FrameAnimationDrawable.this);
                    }
                    break;
                }
            }
        }
    };
    private final Runnable invalidateRunnable = new Runnable() {
        @Override
        public void run() {
            invalidateSelf();
        }
    };
    private boolean autoPlay = true;

    private final Set<WeakReference<Callback>> obtainedCallbacks = new HashSet<>();

    private boolean noMeasure = false;

    public FrameAnimationDrawable(Decoder frameSeqDecoder) {
        paint.setAntiAlias(true);
        this.frameSeqDecoder = frameSeqDecoder;
    }

    public FrameAnimationDrawable(Loader provider) {
        paint.setAntiAlias(true);
        this.frameSeqDecoder = createFrameSeqDecoder(provider, this);
    }

    public void setAutoPlay(boolean autoPlay) {
        this.autoPlay = autoPlay;
    }

    public void setNoMeasure(boolean noMeasure) {
        this.noMeasure = noMeasure;
    }

    protected abstract Decoder createFrameSeqDecoder(Loader streamLoader, BaseFrameSeqDecoder.RenderListener listener);

    /**
     * @param loopLimit <=0为无限播放,>0为实际播放次数
     */
    @SuppressWarnings("unused")
    public void setLoopLimit(int loopLimit) {
        frameSeqDecoder.setLoopLimit(loopLimit);
    }

    public void reset() {
        if (bitmap != null && !bitmap.isRecycled()) {
            bitmap.eraseColor(Color.TRANSPARENT);
        }
        frameSeqDecoder.reset();
    }

    public void pause() {
        frameSeqDecoder.pause();
    }

    public void resume() {
        frameSeqDecoder.resume();
    }

    @SuppressWarnings("unused")
    public boolean isPaused() {
        return frameSeqDecoder.isPaused();
    }

    @Override
    public void start() {
        if (this.frameSeqDecoder.isRunning()) {
            this.frameSeqDecoder.stop();
        }
        this.frameSeqDecoder.reset();
        innerStart();
    }

    private void innerStart() {
        if (BaseFrameSeqDecoder.DEBUG) {
            Log.d(TAG, this + ",start");
        }

        this.frameSeqDecoder.addRenderListener(this);
        if (autoPlay) {
            frameSeqDecoder.start();
        } else {
            if (!this.frameSeqDecoder.isRunning()) {
                this.frameSeqDecoder.start();
            }
        }
    }

    @Override
    public void stop() {
        innerStop();
    }

    private void innerStop() {
        if (BaseFrameSeqDecoder.DEBUG) {
            Log.d(TAG, this + ",stop");
        }

        this.frameSeqDecoder.removeRenderListener(this);
        if (autoPlay) {
            frameSeqDecoder.stop();
        } else {
            this.frameSeqDecoder.stopIfNeeded();
        }
    }

    @Override
    public boolean isRunning() {
        return frameSeqDecoder.isRunning();
    }

    @Override
    public void draw(Canvas canvas) {
        if (bitmap == null || bitmap.isRecycled()) {
            return;
        }
        canvas.setDrawFilter(drawFilter);
        canvas.drawBitmap(bitmap, matrix, paint);
    }

    @Override
    public void setBounds(int left, int top, int right, int bottom) {
        super.setBounds(left, top, right, bottom);
        boolean sampleSizeChanged = frameSeqDecoder.setDesiredSize(getBounds().width(), getBounds().height());
        int sampleSize = frameSeqDecoder.getSampleSize();
        matrix.setScale(
                1.0f * getBounds().width() * sampleSize / frameSeqDecoder.getBounds().width(),
                1.0f * getBounds().height() * sampleSize / frameSeqDecoder.getBounds().height());

        if (sampleSizeChanged)
            this.bitmap = Bitmap.createBitmap(
                    frameSeqDecoder.getBounds().width() / sampleSize,
                    frameSeqDecoder.getBounds().height() / sampleSize,
                    Bitmap.Config.ARGB_8888);
    }

    @Override
    public void setAlpha(int alpha) {
        paint.setAlpha(alpha);
    }

    @Override
    public void setColorFilter(ColorFilter colorFilter) {
        paint.setColorFilter(colorFilter);
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSLUCENT;
    }

    @Override
    public void onStart() {
        Message.obtain(uiHandler, MSG_ANIMATION_START).sendToTarget();
    }

    @Override
    public void onRender(@NonNull ByteBuffer byteBuffer) {
        if (!isRunning()) {
            return;
        }
        if (this.bitmap == null || this.bitmap.isRecycled()) {
            int sampleSize = frameSeqDecoder.getSampleSize();
            this.bitmap = Bitmap.createBitmap(
                    frameSeqDecoder.getBounds().width() / sampleSize,
                    frameSeqDecoder.getBounds().height() / sampleSize,
                    Bitmap.Config.ARGB_8888);
        }
        byteBuffer.rewind();
        if (byteBuffer.remaining() < this.bitmap.getByteCount()) {
            Log.e(TAG, "onRender:Buffer not large enough for pixels");
            return;
        }
        this.bitmap.copyPixelsFromBuffer(byteBuffer);
        uiHandler.post(invalidateRunnable);
    }

    @Override
    public void onEnd() {
        Message.obtain(uiHandler, MSG_ANIMATION_END).sendToTarget();
    }

    @Override
    public boolean setVisible(boolean visible, boolean restart) {
        hookRecordCallbacks();
        if (this.autoPlay) {
            if (BaseFrameSeqDecoder.DEBUG) {
                Log.d(TAG, this + ",visible:" + visible + ",restart:" + restart);
            }
            if (visible) {
                if (!isRunning()) {
                    innerStart();
                }
            } else if (isRunning()) {
                innerStop();
            }
        }
        return super.setVisible(visible, restart);
    }

    @Override
    public int getIntrinsicWidth() {
        if (noMeasure) {
            return -1;
        }
        try {
            return frameSeqDecoder.getBounds().width();
        } catch (Exception exception) {
            return 0;
        }
    }

    @Override
    public int getIntrinsicHeight() {
        if (noMeasure) {
            return -1;
        }
        try {
            return frameSeqDecoder.getBounds().height();
        } catch (Exception exception) {
            return 0;
        }
    }

    @Override
    public void registerAnimationCallback(@NonNull AnimationCallback animationCallback) {
        this.animationCallbacks.add(animationCallback);
    }

    @Override
    public boolean unregisterAnimationCallback(@NonNull AnimationCallback animationCallback) {
        return this.animationCallbacks.remove(animationCallback);
    }

    @Override
    public void clearAnimationCallbacks() {
        this.animationCallbacks.clear();
    }

    public int getMemorySize() {
        int size = frameSeqDecoder.getMemorySize();
        if (bitmap != null && !bitmap.isRecycled()) {
            size += bitmap.getAllocationByteCount();
        }
        return Math.max(1, size);
    }

    @Nullable
    @Override
    public Callback getCallback() {
        return super.getCallback();
    }

    private void hookRecordCallbacks() {
        List<WeakReference<Callback>> lost = new ArrayList<>();
        Callback callback = getCallback();
        boolean recorded = false;
        Set<WeakReference<Callback>> temp = new HashSet<>(obtainedCallbacks);
        for (WeakReference<Callback> ref : temp) {
            Callback cb = ref.get();
            if (cb == null) {
                lost.add(ref);
            } else {
                if (cb == callback) {
                    recorded = true;
                } else {
                    cb.invalidateDrawable(this);
                }
            }
        }
        for (WeakReference<Callback> ref : lost) {
            obtainedCallbacks.remove(ref);
        }
        if (!recorded) {
            obtainedCallbacks.add(new WeakReference<>(callback));
        }
    }

    @Override
    public void invalidateSelf() {
        super.invalidateSelf();
        Set<WeakReference<Callback>> temp = new HashSet<>(obtainedCallbacks);
        for (WeakReference<Callback> ref : temp) {
            Callback callback = ref.get();
            if (callback != null && callback != getCallback()) {
                callback.invalidateDrawable(this);
            }
        }
    }

    public Decoder getFrameSeqDecoder() {
        return frameSeqDecoder;
    }
}

File: frameanimation/src/main/java/com/github/penfeizhou/animation/FrameAnimationDrawable.kt
package com.github.penfeizhou.animation

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.ColorFilter
import android.graphics.DrawFilter
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.PaintFlagsDrawFilter
import android.graphics.PixelFormat
import android.graphics.drawable.Drawable
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.vectordrawable.graphics.drawable.Animatable2Compat
import com.github.penfeizhou.animation.decode.BaseFrameSeqDecoder
import com.github.penfeizhou.animation.decode.FrameSeqDecoder2
import com.github.penfeizhou.animation.decode.RenderListener
import java.lang.ref.WeakReference
import java.nio.ByteBuffer
import kotlin.math.max

abstract class FrameAnimationDrawable(
    val frameSeqDecoder: FrameSeqDecoder2
) : Drawable(), Animatable2Compat {
    private val paint = Paint().apply { isAntiAlias = true }

    private val drawFilter: DrawFilter =
        PaintFlagsDrawFilter(0, Paint.ANTI_ALIAS_FLAG or Paint.FILTER_BITMAP_FLAG)
    private val matrix = Matrix()
    private val animationCallbacks = mutableSetOf<Animatable2Compat.AnimationCallback>()

    private var bitmap: Bitmap? = null
    private val unrecycledBitmap: Bitmap?
        get() = bitmap?.takeUnless { it.isRecycled }

    private val invalidateRunnable = Runnable { invalidateSelf() }
    private var autoPlay = true
    private val obtainedCallbacks: MutableSet<WeakReference<Callback?>> = HashSet()
    private var noMeasure = false

    val isPaused: Boolean
        get() = frameSeqDecoder.isPaused()

    val memorySize: Int
        get() {
            val bitmapSize = unrecycledBitmap?.allocationByteCount ?: 0
            return max(frameSeqDecoder.getMemorySize() + bitmapSize, 1)
        }

    private val renderListener = object : RenderListener {
        private val uiHandler: Handler = Handler(Looper.getMainLooper())
        override fun onStart() {
            for (animationCallback in animationCallbacks.asIterable()) {
                animationCallback.onAnimationStart(this@FrameAnimationDrawable)
            }
        }

        override fun onEnd() {
            uiHandler.post {
                for (animationCallback in animationCallbacks.asIterable()) {
                    animationCallback.onAnimationEnd(this@FrameAnimationDrawable)
                }
            }
        }

        override fun onRender(byteBuffer: ByteBuffer) {
            if (!isRunning) {
                return
            }
            val bitmap = unrecycledBitmap ?: frameSeqDecoder.createBitmap()
            this@FrameAnimationDrawable.bitmap = bitmap

            byteBuffer.rewind()
            if (byteBuffer.remaining() < bitmap.byteCount) {
                Log.e(TAG, "onRender:Buffer not large enough for pixels")
                return
            }
            bitmap.copyPixelsFromBuffer(byteBuffer)
            uiHandler.post(invalidateRunnable)
        }
    }

    init {
        frameSeqDecoder.addRenderListener(renderListener)
    }

    fun setAutoPlay(autoPlay: Boolean) {
        this.autoPlay = autoPlay
    }

    fun setNoMeasure(noMeasure: Boolean) {
        this.noMeasure = noMeasure
    }

    /**
     * @param loopLimit <=0为无限播放,>0为实际播放次数
     */
    fun setLoopLimit(loopLimit: Int) = frameSeqDecoder.setLoopLimit(loopLimit)

    fun reset() {
        unrecycledBitmap?.eraseColor(Color.TRANSPARENT)
        frameSeqDecoder.reset()
    }

    fun pause() = frameSeqDecoder.pause()

    fun resume() = frameSeqDecoder.resume()

    override fun start() {
        if (frameSeqDecoder.isRunning) {
            frameSeqDecoder.stop()
        }
        frameSeqDecoder.reset()
        innerStart()
    }

    private fun innerStart() {
        if (BaseFrameSeqDecoder.DEBUG) {
            Log.d(TAG, "$this,start")
        }
        frameSeqDecoder.addRenderListener(renderListener)
        if (autoPlay || !frameSeqDecoder.isRunning) {
            frameSeqDecoder.start()
        }
    }

    override fun stop() = innerStop()

    private fun innerStop() {
        if (BaseFrameSeqDecoder.DEBUG) {
            Log.d(TAG, "$this,stop")
        }
        frameSeqDecoder.removeRenderListener(renderListener)
        if (autoPlay) {
            frameSeqDecoder.stop()
        } else {
            frameSeqDecoder.stopIfNeeded()
        }
    }

    override fun isRunning(): Boolean = frameSeqDecoder.isRunning

    override fun draw(canvas: Canvas) {
        val bitmap = unrecycledBitmap ?: return
        canvas.drawFilter = drawFilter
        canvas.drawBitmap(bitmap, matrix, paint)
    }

    override fun setBounds(left: Int, top: Int, right: Int, bottom: Int) {
        super.setBounds(left, top, right, bottom)

        val isSampleSizeChanged = frameSeqDecoder.setDesiredSize(bounds.width(), bounds.height())
        val sampleSize = frameSeqDecoder.sampleSize
        val viewport = frameSeqDecoder.getViewport()
        matrix.setScale(
            1.0f * bounds.width() * sampleSize / viewport.width,
            1.0f * bounds.height() * sampleSize / viewport.height
        )
        if (isSampleSizeChanged) {
            bitmap?.recycle()
            bitmap = frameSeqDecoder.createBitmap()
        }
    }

    override fun setAlpha(alpha: Int) {
        paint.alpha = alpha
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
        paint.colorFilter = colorFilter
    }

    @Suppress("DeprecatedCallableAddReplaceWith")
    @Deprecated("Deprecated in Java")
    override fun getOpacity(): Int = PixelFormat.TRANSLUCENT

    override fun setVisible(visible: Boolean, restart: Boolean): Boolean {
        hookRecordCallbacks()
        if (autoPlay) {
            if (BaseFrameSeqDecoder.DEBUG) {
                Log.d(TAG, "$this,visible:$visible,restart:$restart")
            }
            when {
                visible && !isRunning -> innerStart()
                !visible && isRunning -> innerStop()
            }
        }
        return super.setVisible(visible, restart)
    }

    override fun getIntrinsicWidth(): Int =
        if (noMeasure) -1 else frameSeqDecoder.getViewport().width

    override fun getIntrinsicHeight(): Int =
        if (noMeasure) -1 else frameSeqDecoder.getViewport().height

    override fun registerAnimationCallback(animationCallback: Animatable2Compat.AnimationCallback) {
        animationCallbacks.add(animationCallback)
    }

    override fun unregisterAnimationCallback(
        animationCallback: Animatable2Compat.AnimationCallback
    ): Boolean = animationCallbacks.remove(animationCallback)

    override fun clearAnimationCallbacks() = animationCallbacks.clear()

    private fun hookRecordCallbacks() {
        val callback = callback

        obtainedCallbacks.removeAll { it.get() == null }
        for (ref in obtainedCallbacks.toSet()) {
            ref.get()
                ?.takeIf { it != callback }
                ?.invalidateDrawable(this)
        }

        if (!obtainedCallbacks.any { it.get() == callback }) {
            obtainedCallbacks.add(WeakReference(callback))
        }
    }

    override fun invalidateSelf() {
        super.invalidateSelf()
        for (ref in obtainedCallbacks) {
            val callback = ref.get()?.takeIf { it != callback }
            callback?.invalidateDrawable(this)
        }
    }

    companion object {
        private val TAG = FrameAnimationDrawable::class.java.simpleName

        private fun FrameSeqDecoder2.createBitmap(): Bitmap {
            val viewport = getViewport()
            return Bitmap.createBitmap(
                viewport.width / sampleSize,
                viewport.height / sampleSize,
                Bitmap.Config.ARGB_8888
            )
        }
    }
}
